### PR TITLE
add strike selection changes

### PR DESCRIPTION
**Pull request by chuddster**

### CODE REVIEW

File: contracts/utils/StrikeSelection.sol
Comment (chuddster) [inline contracts/utils/StrikeSelection.sol:44]: decimals here not entirely accurate - will adjust when I run tests

File: contracts/utils/StrikeSelection.sol
Comment (kenchangh) [inline contracts/utils/StrikeSelection.sol:62]: Love the approach on how to get strike price! A few comments:

- We should bound the number of iterations with a for loop instead of using a while loop. I haven't verified but it seems possible end up in an infinite loop
- Maybe fallback to whatever's the best strike price?

File: contracts/utils/StrikeSelection.sol
Comment (kenchangh) [inline contracts/utils/StrikeSelection.sol:73]: One desirable thing we should aim for is to select strike prices that have similar increments (+100 for ETH, +1000 for BTC) as Deribit, this allows people to easily hedge on Deribit

image

The loop should iterate through the list with increments of +100: 3000, 3100, 3200, 3300

File: contracts/utils/StrikeSelection.sol
Comment (chuddster) [reply contracts/utils/StrikeSelection.sol:62]: So as it stands now, the delta will always be decreasing for calls and it will be increasing for puts. For calls we are looking for 0.1d and for puts 0.9d. When we iterate at each step we are either increasing or decreasing the delta. At some point we will reach the threshold of 0.1 and 0.9 and currDelta > newDelta will be true for puts or currDelta < newDelta will be true for calls.

We can add logic to just do % from spot price if while loop goes past x iterations? the only way this is possible is if chainlink asset oracle fails and latestAnswer() gives a wacky answer

File: contracts/utils/StrikeSelection.sol
Comment (chuddster) [reply contracts/utils/StrikeSelection.sol:73]: so do you want to just make step an absolute term that can be manually changed? ex: setStep -> 100 for eth / setStep -> 1000. then if eth appreciates we might do setStep -> 200 for eth based on what deribit does

Comment (kenchangh) [reply contracts/utils/StrikeSelection.sol:73]: Will leave that up to you, as long as we can minimize the iterations and gas use, I'm ok with either approach

File: contracts/utils/StrikeSelection.sol
Comment (kenchangh) [reply contracts/utils/StrikeSelection.sol:62]: Thanks for the clarification, makes sense that the iteration is bounded even with a while loop because it's just monotonically increasing/decreasing

File: contracts/utils/StrikeSelection.sol
Comment (kenchangh) [inline contracts/utils/StrikeSelection.sol:46]: Would be helpful to return `(strikePrice, delta)` so the consumer can emit the delta value out in an event for transparency

Comment (chuddster) [reply contracts/utils/StrikeSelection.sol:46]: 👍

File: contracts/utils/StrikeSelection.sol
Comment (chuddster) [reply contracts/utils/StrikeSelection.sol:70]: Yeah...think it makes sense to leave as is. Even if we do in absolute terms will still be imprecise (ex current price is 2374 and we have increments of 100 will be 2474, 2574 so wont match perfectly either way.) Can always modulate iterations by increasing step

Comment (kenchangh) [reply contracts/utils/StrikeSelection.sol:77]: If we iterate to 2474, we can always do a modulo so that it resolves correctly to 2400

`strike = strike - (strike % 100)`

File: test/StrikeSelection.ts
Comment (kenchangh) [inline test/StrikeSelection.ts:104]: Add test for tie breaks

File: contracts/utils/StrikeSelection.sol
Comment (kenchangh) [inline contracts/utils/StrikeSelection.sol:71]: One easy fix to round out the strike prices to be like Deribit is to modulo the starting `strike`

```solidity
uint256 strike = assetPrice.sub(assetPrice % step);
```

So if we read 2742 as the `assetPrice`, we would set `strike = 2700` and iterate +-100 from there

File: contracts/utils/StrikeSelection.sol
Comment (kenchangh) [inline contracts/utils/StrikeSelection.sol:95]: Before we return, we should do a `require(isPut ? strike < assetPrice : strike > assetPrice)` so we have a final sanity check

**Review by kenchangh — DISMISSED** [review]
Changes LGTM, let's fix the tests before merging

**Review by kenchangh — APPROVED** [review]

**Pull request merged**
